### ISSUE TITLE
How to get the model file with *.cpb or *.pba.z

### ISSUE THREAD

**Issue by lyq123180** [elucideye/xgboostcpp#15]
Finally I successful compile the project in Android NDK r14b.
I train the xgboost model using the python API and get the binary model file.
I try to load the model with xgboostcpp using the "load_model" function,but failed.
I need the model file  with *.cpb or *.pba.z format.
How I convert my model from python API to this format?
Thanks!

**Comment by lyq123180** [comment]
I notice your answer in this page
https://github.com/dmlc/xgboost/issues/1016
Now I know the .cpb file is the cereal portable binary archive format.
Still,I don't know how to convert the model file.
You know training a model and tuning the parameters in C API is suffering.
I scan the Document of cereal and can't find anything.

**Comment by headupinclouds** [comment]
> I need the model file with *.cpb or *.pba.z format.  How I convert my model from python API to this format?

Due to added complexity, I dropped the *.pba.z boost models.  Only the *.cpb models are currently supported.  These are not compatible with or easily convertible to/from the standard xgboost models.  Note that this code is also based on a modified xgboost v0.4.0 fork, and the current xgboost version is somewhere >= v0.6.0.  This sample code is intended to support self consistent training and testing using the provided C++ API layer only.  

This CMake option can be used to disable the cereal archives, and it should default to the standard v0.4.0 (caveat: I haven't tested it that way recently):
```
option(XGBOOSTER_SERIALIZE_WITH_CEREAL "Serialize w/ cereal" ON) # hunter update
```

For typical use you would would first train a model using this application (modified as needed for your application):
xgboostcpp/src/test/test-xgbooster.cpp

Then you would link the exported library in your android code:
xgboostcpp/src/lib/xgboostcpp

You could then adapt the prediction part of the test-xgbooster code into your Android project as a starting point.  This has worked well for me for regression problems in iOS and Android projects.  It only supports the subset of xgboost regression parameters that I've needed for my application, but can easily by extended.

To use that training code for your particular problem, you will probably want to modify the xgboost parameters, which could be done in the `xgboostcpp::XGBooster::Recipe`  block of code shown below (a subset of current xgboost parameters):

https://github.com/elucideye/xgboostcpp/blob/1021f11ab203403e53b35c99533ef12ffcecb95f/src/test/test-xgbooster.cpp#L62-L67

The end-to-end worfkflow, using https://github.com/ruslo/polly, would look something like this:

1)  build the training code on a host platform w/ a suitable toolchain :

```
polly.py --toolchain libcxx --install --reconfig --verbose
```

2) train from *.csv files using `test-xgbooster` modified as needed for your application

3) build the library for use on Android platforms with something like this (or another suitable toolchain):

```
polly.py --toolchain android-ndk-r14b-api-21-armeabi-v7a-neon-clang-libcxx.cmake --install --reconfig --verbose
```

This is just some sample code made available in case it is useful for other people dealing with cross platform + mobile requirements w/ the C++ API.

**Comment by lyq123180** [comment]
Thanks for the careful reply  ：）
I prepared to train my model in C API.
But my task is multiclass softmax classification，I notice In Recipe there is a regression enable .
can i make multiclass classification with disable the regression？

**Comment by headupinclouds** [comment]
> I notice In Recipe there is a regression enable 

That was all I needed, but it should be fairly straightforward to add hooks for classification mode as well.

> But my task is multiclass softmax classification，

We would need to add a new enumerated type in the `Recipe` structure to control the xgboost `objective` parameter, and for softmax we would need to add a `vector<float>` output to the API.

If you have a reasonably compact softmax test set, we could add a new unit test similar to the one here:

https://github.com/elucideye/xgboostcpp/blob/master/src/lib/xgboostcpp/ut/test-xgboostcpp-regression.cpp

Feel free to add a PR for this.  I can help with it, but probably not much this week.
